rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,509 @@
+
+- centralize the logging system in the coordinator (and have the workers send messages to the coordinator when using the logging system)
+
+We could go further, and change the way the logging system works in clustering mode by having the coordinator centralize the logging system. The worker’s logger implementation would just send messages to the coordinator. If this may be a correct design, the main downside is that the logging implementation would be totally different in cluster and non cluster mode, and is overall way more work that the other options.
+
+Overall, if no option is trivial, I feel like option 1) is still the most pragmatic one. Option 3) is probably a better 
+design, but represents more work. It would probably be fine if all our appenders were impacted, but as only the 
+rolling-file one is, I feel like going with 1) would be ok.
+
+### The status API
+
+In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
+while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
+meaning that it would not be possible to know the status of the cluster as a whole.
+
+We will need to add a centralized status service in the coordinator. Also, as the `/status` endpoint cannot be served 
+from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
+the status endpoint.
+
+We may also want to have the `/status` endpoint display each individual worker status in addition to the 
+global status, which may be breaking change in the `/status` API response format.
+
+### PID file
+
+Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 
+usage, as the PID stored in the file will be a arbitrary worker’s PID, instead of the coordinator (main process) PID.
+
+In clustering mode, we will need to have to coordinator handle the PID file logic, and to disable pid file handling 
+in the worker's environment service.
+
+### SavedObjects migration
+
+In the current state, all workers are going to try to perform the migration. Ideally, we would have only one process 
+perform the migration, and the other ones just wait for a ready signal. We can’t easily have the coordinator do it, 
+so we would probably have to leverage the ‘main worker’ concept to do so.
+
+The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
+Kibana, so this can probably be considered an improvement.
+
+## Open questions and things to solve
+
+### Memory consumption
+In cluster mode, node options such as `max-old-space-size` will be used by all processes. E.g using 
+`--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 
+
+If this something we will need to document somewhere?
```

Comment:
> Clustering should probably be disabled by default

Yea, it will likely be the case

> In the clustering docs we could explain all the implications of scaling out Kibana.

Yea. even without using any memory option, we still need to document that we clustering mode will spawn multiple processes, with all the underlying implications.